### PR TITLE
Add support for saved model in 2.0

### PR DESCRIPTION
**Pull request by dfalbel**
This PR uses:

* [`tf.saved_model.save`](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/saved_model/save) to back the  `export_savedmodel` method for keras models. It uses `...` to pass aditional arguments.

Implements `model_to_saved_model` and `model_from_saved_model` wraping [`tf.keras.experimental.export_saved_model`](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/experimental/export_saved_model) and [`tf.keras.experimental.load_saved_model`](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/experimental/load_from_saved_model).

The main difference of this functions and `export_savedmodel` is that it allows you to load the model as a usual Keras model and keep training the model, while using `export_savedmodel` doesn't provide this helpers.

Added `save_model_weights_tf` and `load_model_weights_tf` to save and load weights in the TensorFlow SavedModel format. This is now available via the `save_format` argument of the [`save_weights`](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/Model#save_weights) method.

Also added a vignette explaining all different serializing options. Rendered [here](http://rpubs.com/dfalbel/saving_serializing)

### CODE REVIEW

**Comment by dfalbel** [comment]
@skeydan Can you take a look?

File: vignettes/saving_serializing.Rmd [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+> This tutorial is an R translation of [this page](https://www.tensorflow.org/alpha/guide/keras/saving_and_serializing)
+available in the official TensorFlow documentation.
+
+The first part of this guide covers saving and serialization for Sequential models and 
+models built using the Functional API and for Sequential models. The saving and serialization
```
Comment (skeydan) [inline vignettes/saving_serializing.Rmd]: I know this is copied from TF but I don't get what it says :-)

File: vignettes/saving_serializing.Rmd [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+
+## Saving Sequential Models or Functional models
+
+```{r}
+inputs <- layer_input(shape = 784, name = "digits")
+outputs <- inputs %>% 
+  layer_dense(units = 64, activation = "relu", name = "dense_1") %>% 
+  layer_dense(units = 64, activation = "relu", name = "dense_2") %>% 
+  layer_dense(units = 10, activation = "softmax", name = "predictions")
+model <- keras_model(inputs, outputs) # can't name the model
```
Comment (skeydan) [inline vignettes/saving_serializing.Rmd]: can't name means, from our side? If so, should we make this more explicit and/or investigate? Or do we just need to add a parameter?

File: vignettes/saving_serializing.Rmd [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+
+```{r}
+c(c(x_train, y_train), c(x_test, y_test)) %<-% dataset_mnist()
+x_train <- x_train %>% array_reshape(dim = c(60000, 784))/255
+x_test <- x_test %>% array_reshape(dim = c(10000, 784))/255
+
+model %>% compile(loss = "sparse_categorical_crossentropy",
+                  optimizer = optimizer_rmsprop())
+
+history <- model %>% fit(x_train, y_train, btach_size = 64, epochs = 1)
```
Comment (skeydan) [inline vignettes/saving_serializing.Rmd]: typo batch_size

File: vignettes/saving_serializing.Rmd [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+* A TensorFlow checkpoint containing the model weights.
+* A SavedModel proto containing the underlying Tensorflow graph. Separate graphs are saved for prediction (serving), train, and evaluation. If the model wasn't compiled before, then only the inference graph gets exported.
+* The model's architecture config, if available.
+
+### Architecture-only saving
+
+Sometimes, you are only interested in the architecture of the model, and you 
+don't need to save the weight values or the optimizer. In this case, you can 
+retrieve the "config" of the model via the get_config() method. The config is 
+a Python dict that enables you to recreate the same model -- initialized from
```
Comment (skeydan) [inline vignettes/saving_serializing.Rmd]: named list, for us?

File: vignettes/saving_serializing.Rmd [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+```
+
+```{r}
+# Note that the model state is not preserved! We only saved the architecture.
+new_predictions <- predict(reinitialized_model, x_test)
+all.equal(predictions, new_predictions)
+```
+
+You can alternatively use `model_to_json()` and `model_from_json()`, which uses a 
+JSON string to store the config instead of a Python dict. This is useeful to save
```
Comment (skeydan) [inline vignettes/saving_serializing.Rmd]: useful (typo)

File: vignettes/saving_serializing.Rmd [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+Note that save_weights can create files either in the Keras HDF5 format, or in 
+the TensorFlow SavedModel format.
+
+```{r}
+save_model_weights_tf(model, "model_weights")
+```
+
+## Saving Subclassed Models
+
+Sequential models and Functional models are datastructures that represent a
```
Comment (skeydan) [inline vignettes/saving_serializing.Rmd]: data structures

File: vignettes/saving_serializing.Rmd [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+```{r}
+save_model_weights_tf(model, "model_weights")
+```
+
+## Saving Subclassed Models
+
+Sequential models and Functional models are datastructures that represent a 
+DAG of layers. As such, they can be safely serialized and deserialized.
+
+A subclassed model differs in that it's not a datastructure, it's
```
Comment (skeydan) [inline vignettes/saving_serializing.Rmd]: data structure

File: vignettes/saving_serializing.Rmd [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+new_model <- keras_model_simple_mlp(num_classes = 10)
+new_model %>% compile(loss = "sparse_categorical_crossentropy",
+                  optimizer = optimizer_rmsprop())
+
+# This initializes the variables used by the optimizers,
+# as well as any stateful metric variables
+train_on_batch(new_model, x_train[1:5,], y_train[1:5])
+
+# Load the state of the old model
+# add a load_weights fun?
```
Comment (skeydan) [inline vignettes/saving_serializing.Rmd]: is this line obsolete?

File: vignettes/saving_serializing.Rmd (hunk 2 above)
Comment (dfalbel) [reply vignettes/saving_serializing.Rmd]: Yes, it's from our side. I'll open another PR for this

**Comment by dfalbel** [comment]
Thanks for the review! 

In theory, both `saved_model_weights` functions work with Custom Models and Sequential/Functional models. That's why I kept the same names. Do you think it's reasonable?

The tricky part with custom models is that you need to run a dummy training step just to initialize the weights. 

In `tf.keras` when you use `model.save_weights(fpath)` it will use the SavedModel format unless `fpath` has an `.h5` or `.keras` extension. 

It should work with your eager posts if you are using `keras` custom models!

**Comment by skeydan** [comment]
> In theory, both saved_model_weights functions work with Custom Models and Sequential/Functional models. That's why I kept the same names. Do you think it's reasonable?

Yeah what I meant was that I like how effectively, you're distinguishing between both of them by adding that `_tf` suffix in the one case...

> The tricky part with custom models is that you need to run a dummy training step just to initialize the weights.

Interestingly, I think that in the meantime, I encountered other cases where models have not been built. I don't recall right now, but it could be the same when using `eager` with `Sequential` models (((and no `input` layer specified)))

File: vignettes/saving_serializing.Rmd [hunk 9]
```diff
@@ -297,7 +317,7 @@ associated with the model: - The layers' weights - The optimizer's state - Any v
 associated with stateful model metrics (if any).
 
 ```{r}
-save_model_weights_tf(model, "my_weights")
+save_model_weights_hdf5(model, "my_weights.h5")
```
Comment (skeydan) [inline vignettes/saving_serializing.Rmd]: Now I'm confused... didn't we say

> The recommended way to save a subclassed model is to use `save_model_weights_tf` to 
create a TensorFlow SavedModel checkpoint, which will contain the value of all variables 
associated with the model:

?

Comment (dfalbel) [reply vignettes/saving_serializing.Rmd]: Oh sorry, I change it to test if it worked and forgot to change back :)

Comment (skeydan) [reply vignettes/saving_serializing.Rmd]: ah okay, I'm relieved it wasn't some fundamental misunderstanding from my side ;-))
... and does it work (I guess not)?
